test/rulesets/Base/array.jl

```diff
@@ -1,3 +1,14 @@
+@testset "constructors" begin
+    @testset "undef" begin
+        test_rrule(Array{Float64, 1}, undef, 5)
+        test_rrule(Array{Float32, 3}, undef, 5, 4, 3)
+    end
+    @testset "from existing array" begin
+        test_rrule(Array, randn(2, 5))
+        test_rrule(Array, Diagonal(randn(5)))
+    end
```

Comment:
I haven't tried these, but they appear to be allowed by the rule:
```suggestion
        test_rrule(Matrix, Diagonal(randn(5)))
        test_rrule(Matrix, transpose(randn(4)))
        test_rrule(Array{ComplexF64}, randn(3))
    end
```

Reply:
Seems reasonable -- thanks for the extra tests.